rest-client/src/main/java/org/sagebionetworks/bridge/rest/ErrorResponseInterceptor.java

```diff
@@ -72,41 +74,46 @@ private void throwExceptionOnErrorStatus(String url, int statusCode, JsonElement
                 message = "There has been an error on the server";
             }
         }
-
-        BridgeSDKException e;
-        if (statusCode == 401) {
-            e = new NotAuthenticatedException(message, url);
-        } else if (statusCode == 403) {
-            e = new UnauthorizedException(message, url);
-        } else if (statusCode == 404) {
-            e = new EntityNotFoundException(message, url);
-        } else if (statusCode == 410) {
-            e = new UnsupportedVersionException(message, url);
-        } else if (statusCode == 412) {
+        // This does not return an exception message, it returns session object.
+        if (statusCode == 412) {
             UserSessionInfo session = null;
             if (node != null) {
                 session = RestUtils.GSON.fromJson(node, UserSessionInfo.class);
             }
-            e = new ConsentRequiredException("Consent required.", url, session);
-        } else if (statusCode == 409 && message.contains("already exists")) {
-            e = new EntityAlreadyExistsException(message, url);
-        } else if (statusCode == 409 && message.contains("has the wrong version number")) {
-            e = new ConcurrentModificationException(message, url);
-        } else if (statusCode == 400 && message.contains("A published survey")) {
-            e = new PublishedSurveyException(message, url);
-        } else if (statusCode == 400) {
-            if (node != null && node.getAsJsonObject().get("errors") != null) {
-                Map<String, List<String>> errors = RestUtils.GSON.fromJson(node.getAsJsonObject().get("errors"),
-                        ERRORS_MAP_TYPE_TOKEN);
-                e = new InvalidEntityException(message, errors, url);
-            } else {
-                e = new BadRequestException(message, url);
-            }
-        } else if (statusCode == 501) {
-            e = new NotImplementedException(message, url);
-        } else {
-            e = new BridgeSDKException(message, statusCode, url);
+            throw new ConsentRequiredException("Consent required.", url, session);
+        }
+        String type = (node != null) ? node.getAsJsonObject().get("type").getAsString() : "Unknown";
+        switch(type) {
+            case "NotAuthenticatedException":
+                throw new NotAuthenticatedException(message, url);
```

Comment:
Also, how much work would it be to use Reflection to handle most of these use cases. That way, we don't need a big switch statement to convert an exception name to an exception class with otherwise identical constructor invocations.